### PR TITLE
fix: gracefully handle process kill

### LINKED ISSUE #406: [binary] Stop service doesn't kill the process
Start, then stop a service
You can check processes with `lsof -i -P -n | grep LISTEN`

### LINKED ISSUE #437: On Quit kills running services
we should catch Quit events and kill all running services

### PR DESCRIPTION
**Pull request by swarnimarun**
Fixes #406 

---

### Why was stop service not working properly?

SIGKILL is sent by `std::process::Child::kill(&self)` and the FastAPI server is hosted in a child process, hence killing the parent just causes a reparent to current process, and doesn't kill the child process spawned by Python. While killing the prem-app cleans them up.
This is cause, the process groups aren't initialized for `std::process::Command` for grouped signal dispatch.

### Possible Solutions

There are two solutions,
1. Use process groups for each command (CommandGroup?), to broadcast SIGKILL command to the entire process tree for the quiting command.
2. Use SIGTERM instead for then the process will be responsible to handle the resource cleanup, this also ensures no broken resources will be left behind that could be removed by the process.

#### We picked solution 2.

---

Fixes #437 

---

### Ensure on quit operation we gracefully stop all services

This should already be handled because the terminal session will start it in a process group that ensure all the process will be killed if they haven't been detached but currently it doesn't guarantee deterministic behaviour. 

### Possible Solutions

There are two solutions,
1. Before application quit call stop_all_services.
2. Before binary exit or panic of any kind try to call stop_all_services.

### Cons

Because we handle panics and exits with custom code that code itself "can" "theoritically" panic and cause a panic during a panic, which means a direct abort leading to no stack trace being reported.
This should currently not happen as the code called in such contexts is panic-free, and hence should be safe behaviour but letting people know just in case we ever introduce a panic to the code, it should be easier to debug.

### CODE REVIEW

File: src-tauri/src/controller_binaries.rs [hunk 1]
```diff
@@ -158,30 +167,62 @@ pub async fn start_service(
...
-                log::error!("Failed to kill child process: {}", e);
-            }
-        }
+pub async fn stop_service(service_id: String, state: State<'_, Arc<SharedState>>) -> Result<()> {
+    let running_services = &state.running_services;
+    let services = &state.services;
+    _stop_service(service_id.as_str(), running_services, services).await
+}
+
+pub async fn _stop_service(
```
Comment (tiero) [inline src-tauri/src/controller_binaries.rs:176]: shall we keep it `pub` if is internal method?

File: src-tauri/src/utils.rs [hunk 2]
```diff
@@ -26,21 +32,19 @@ pub fn is_x86_64() -> bool {
 }
 
 pub fn get_binary_url(binaries_url: &HashMap<String, Option<String>>) -> Result<String> {
-    let mut binary_url = "".to_string();
     if is_aarch64() {
-        binary_url = binaries_url
+        binaries_url
             .get("aarch64-apple-darwin")
-            .unwrap()
+            .with_context(|| "No binary available for platform")?
```
Comment (tiero) [inline src-tauri/src/utils.rs:38]:
```suggestion
            .with_context(|| "Service not supported on your platform")?
```

File: src-tauri/src/controller_binaries.rs (hunk 1 above)
Comment (swarnimarun) [reply src-tauri/src/controller_binaries.rs:176]: yeah that seems to have been left in unnecessarily.

**Comment by Janaka-Steph** [comment]
When service Mistral Orca is running, calling restart causes error `Service with `mistral-7b-openorca` already exist`. 
It seems that `running_services` is not cleared properly.

**Review by Janaka-Steph — COMMENTED** [review]
Quitting the app from top menu bar make the app unresponsive and doesn't quit.
Screenshot 2023-10-30 at 10 49 36


Also when service Mistral Orca is running, calling restart causes error "Service with mistral-7b-openorca already exist".
It seems that running_services is not cleared properly.

File: src-tauri/src/main.rs [hunk 3]
```diff
@@ -164,11 +186,11 @@ fn main() {
 
     let system_tray = SystemTray::new().with_menu(tray_menu);
 
-    let state = SharedState::default();
-    #[allow(unused_mut)]
-    let mut app = tauri::Builder::default()
+    let state = std::sync::Arc::new(SharedState::default());
```
Comment (Janaka-Steph) [inline src-tauri/src/main.rs:189]: Can you explain why Arc is necessary?

**Pull request merged**